ton_client/client/src/client/wasm_client_env.rs

```diff
@@ -0,0 +1,318 @@
+use crate::error::ClientResult;
+use super::{Error, WebSocket, FetchMethod, FetchResult};
+
+fn js_value_to_value(js_value: JsValue) -> ClientResult<serde_json::Value> {
+    js_value.into_serde::<serde_json::Value>()
+        .map_err(|_| Error::cannot_convert_jsvalue_to_json(js_value))
+}
+
+fn js_value_to_string(js_value: JsValue) -> String {
+    if let Ok(txt) = js_value.clone().dyn_into::<js_sys::Error>() {
+        String::from(txt.message())
+    } else {
+        js_value_to_value(js_value)
+            .map(|val| format!("{:#}", val))
+            .unwrap_or("Unserializable value".to_owned())
+    }
+}
+
+// web-sys and wasm-bindgen types are not `Send` so we cannot use them directly in async
+// functions which are registered in dispather: registration requires `Future` returned
+// by the function to be `Send`, but using not `Send` types inside prevents it. So we have
+// to process functions in another task which encapsulates work with non-`Send` types
+async fn execute_spawned<R, Fut, F>(func: F) -> ClientResult<R>
+where
+    R: 'static,
+    Fut: Future<Output=R> + 'static,
+    F: FnOnce() -> Fut + 'static
+{
+    let (sender, receiver) = tokio::sync::oneshot::channel();
+
+    wasm_bindgen_futures::spawn_local(async move {
+        let _ = sender.send(func().await);
+    });
+
+    receiver
+        .await
+        .map_err(|err| Error::can_not_receive_spawned_result(err))
+}
+
+pub(crate) struct ClientEnv {}
+
+impl ClientEnv {
+    pub fn new() -> ClientResult<Self> {
+        Ok(Self {})
+    }
+
+    /// Sets timer for provided time interval
+    pub async fn set_timer_internal(ms: u64) -> ClientResult<()> {
+        let window = web_sys::window()
+            .ok_or_else(|| Error::set_timer_error("Can not get `window`"))?;
+
+        let (sender, receiver) = tokio::sync::oneshot::channel();
+        let on_timer = Closure::once(move || {
+            let _ = sender.send(());
+        });
+
+        window
+            .set_timeout_with_callback_and_timeout_and_arguments_0(
+                on_timer.as_ref().unchecked_ref(),
+                ms as i32
+            ).map_err(|_| Error::set_timer_error("Can not set timer"))?;
+
+        receiver
+            .await
+            .map_err(|_| Error::set_timer_error("Can not receiver timer result"))?;
+
+        Ok(())
+    }
+
+    /// Connects to the websocket endpoint
+    pub async fn websocket_connect_internal(
+        url: &str,
+        headers: Option<HashMap<String, String>>,
+    ) -> ClientResult<WebSocket> {
+        // Connect to a server
+        let ws = if let Some(Some(protocols)) = headers.map(|mut headers| headers.remove("Sec-WebSocket-Protocol")) {
+            web_sys::WebSocket::new_with_str(url, &protocols)
+        } else {
+            web_sys::WebSocket::new(url)
+        }
+        .map_err(|_| Error::websocket_connect_error(url, "cannot create websocket"))?;
+
+        let (receive_sink, receive_stream) = futures::channel::mpsc::channel(100);

```

Comment:
suggest to rename `on_message_sink`, `on_message_stream`

Reply:
Done